File: backend/app/schemas/campaign_optimizer.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any

class SalesCategoryTrendDTO(BaseModel):
    category: str
    historical_revenue: float
    order_volume: int
    yoy_growth_pct: float
    avg_margin_pct: float
    sales_velocity: str  # "HIGH", "MODERATE", "ACCELERATING", "DECLINING"
    top_selling_sku: str

class CampaignOpportunityDTO(BaseModel):
    id: str
    title: str
    category: str
    target_segment: str
    rationale: str
    recommended_discount_pct: float
    estimated_revenue_potential: float
    confidence_score: float
    target_skus: List[str]

class DailyTrajectoryPointDTO(BaseModel):
    day: int
    date_label: str
    baseline_revenue: float
    campaign_revenue: float
    incremental_lift: float
    conversion_rate_pct: float
    orders_count: int

class ChannelPerformanceDTO(BaseModel):
    channel: str  # "WhatsApp Business", "Email", "In-App Banner", "SMS Gateway"
    sent_count: int
    open_rate_pct: float
    click_through_rate_pct: float
    conversion_rate_pct: float
    revenue_generated: float
    roas: float

class CampaignImprovementDTO(BaseModel):
    id: str
    campaign_id: str
    campaign_name: str
    recommendation_type: str  # "CHANNEL_REALLOCATION", "DISCOUNT_OPTIMIZATION", "MIN_ORDER_TWEAK", "BUNDLE_ATTACHMENT", "PACING_ACCELERATION"
    insight: str
    recommended_improvement: str
    expected_additional_lift: str
    expected_lift_inr: float
    confidence_score: float
    status: str = "PENDING"  # "PENDING" | "APPLIED"
    applied_at: Optional[str] = None

class OptimizedCampaignDTO(BaseModel):
    id: str
    name: str
    # Required by User Specification:
    target_products: List[str] = Field(..., description="List of target product names or SKUs")
    campaign_objective: str = Field(..., description="Strategic campaign objective")
    predicted_roi: float = Field(..., description="Predicted Return on Investment percentage e.g. 340.0")
    predicted_roi_display: str
    estimated_revenue_increase: float = Field(..., description="Estimated net revenue increase in INR")
    estimated_revenue_increase_display: str
    confidence_score: float = Field(..., ge=0.0, le=1.0, description="Confidence score between 0.0 and 1.0")
    
    # Suggested discount & optimization details:
    suggested_discount_pct: float = Field(..., description="AI suggested optimal discount percentage")
    discount_code: str
    target_segment: str
    target_segment_id: str
    status: str = "active"  # "active" | "scheduled" | "completed" | "draft"
    channels: List[str] = ["WhatsApp Business", "Email"]
    
    # Financial metrics:
    campaign_spend_budget: float
    baseline_revenue: float
    projected_total_gmv: float
    projected_orders: int
    conversion_lift_pct: float
    price_elasticity: float
    net_margin_impact_pct: float
    
    # Creative & Timing:
    ai_copy_subject: str
    ai_copy_body: str
    start_date: str
    end_date: str
    duration_days: int
    created_at: str
    
    # Tracking & Telemetry:
    trajectory: List[DailyTrajectoryPointDTO] = []
    channel_performance: List[ChannelPerformanceDTO] = []

class GenerateOptimizedCampaignRequestDTO(BaseModel):
    opportunity_id: Optional[str] = None
    target_segment_id: str
    campaign_objective: str
    target_products: List[str]
    suggested_discount_pct: Optional[float] = 15.0
    min_order_value: Optional[float] = 5000.0
    duration_days: Optional[int] = 14
    channels: Optional[List[str]] = ["WhatsApp Business", "Email"]

class CampaignOptimizerOverviewDTO(BaseModel):
    total_campaigns: int
    active_campaigns: int
    total_projected_revenue_increase: float
    avg_predicted_roi: float
    avg_confidence_score: float
    top_performing_channel: str
    historical_sales_trends: List[SalesCategoryTrendDTO]
    identified_opportunities: List[CampaignOpportunityDTO]
    campaigns: List[OptimizedCampaignDTO]
    active_improvements: List[CampaignImprovementDTO]
    channel_attribution_summary: List[ChannelPerformanceDTO]

File: frontend/src/types/campaign-optimizer.ts
export interface SalesCategoryTrend {
  category: string;
  historical_revenue: number;
  order_volume: number;
  yoy_growth_pct: number;
  avg_margin_pct: number;
  sales_velocity: 'HIGH' | 'MODERATE' | 'ACCELERATING' | 'DECLINING';
  top_selling_sku: string;
}

export interface CampaignOpportunity {
  id: string;
  title: string;
  category: string;
  target_segment: string;
  rationale: string;
  recommended_discount_pct: number;
  estimated_revenue_potential: number;
  confidence_score: number;
  target_skus: string[];
}

export interface DailyTrajectoryPoint {
  day: number;
  date_label: string;
  baseline_revenue: number;
  campaign_revenue: number;
  incremental_lift: number;
  conversion_rate_pct: number;
  orders_count: number;
}

export interface ChannelPerformance {
  channel: string;
  sent_count: number;
  open_rate_pct: number;
  click_through_rate_pct: number;
  conversion_rate_pct: number;
  revenue_generated: number;
  roas: number;
}

export interface CampaignImprovement {
  id: string;
  campaign_id: string;
  campaign_name: string;
  recommendation_type: string;
  insight: string;
  recommended_improvement: string;
  expected_additional_lift: string;
  expected_lift_inr: number;
  confidence_score: number;
  status: 'PENDING' | 'APPLIED';
  applied_at?: string;
}

export interface OptimizedCampaign {
  id: string;
  name: string;
  target_products: string[];
  campaign_objective: string;
  predicted_roi: number;
  predicted_roi_display: string;
  estimated_revenue_increase: number;
  estimated_revenue_increase_display: string;
  confidence_score: number;
  suggested_discount_pct: number;
  discount_code: string;
  target_segment: string;
  target_segment_id: string;
  status: 'active' | 'scheduled' | 'completed' | 'draft';
  channels: string[];
  campaign_spend_budget: number;
  baseline_revenue: number;
  projected_total_gmv: number;
  projected_orders: number;
  conversion_lift_pct: number;
  price_elasticity: number;
  net_margin_impact_pct: number;
  ai_copy_subject: string;
  ai_copy_body: string;
  start_date: string;
  end_date: string;
  duration_days: number;
  created_at: string;
  trajectory: DailyTrajectoryPoint[];
  channel_performance: ChannelPerformance[];
}

export interface CampaignOptimizerOverview {
  total_campaigns: number;
  active_campaigns: number;
  total_projected_revenue_increase: number;
  avg_predicted_roi: number;
  avg_confidence_score: number;
  top_performing_channel: string;
  historical_sales_trends: SalesCategoryTrend[];
  identified_opportunities: CampaignOpportunity[];
  campaigns: OptimizedCampaign[];
  active_improvements: CampaignImprovement[];
  channel_attribution_summary: ChannelPerformance[];
}
